From 151143110af565eb094d0eddcb2a3d5e7af20cc9 Mon Sep 17 00:00:00 2001
From: Evan Huus <eapache@gmail.com>
Date: Wed, 26 Feb 2014 13:00:03 -0500
Subject: [PATCH 2/3] Don't mix emem and glib memory and init routines.

The seasonal memory is freed before the init routine is called, leading to a
whole bunch of use-after-free errors.

Fixes bug #9802 (and duplicates).

This introduces a few minor leaks but I can't find an easy way to add additional
free calls that doesn't lead to double-free errors.

Change-Id: I1536fcb8e96f5560ad366169e815f62967b3e40d
Reviewed-on: https://code.wireshark.org/review/407
Reviewed-by: Evan Huus <eapache@gmail.com>
---
 epan/dissectors/packet-rlc.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/epan/dissectors/packet-rlc.c b/epan/dissectors/packet-rlc.c
index 86e2c07..dbeb06a 100644
--- a/epan/dissectors/packet-rlc.c
+++ b/epan/dissectors/packet-rlc.c
@@ -357,7 +357,7 @@ rlc_sdu_create(void)
 {
 	struct rlc_sdu *sdu;
 
-	sdu = se_alloc0(sizeof(struct rlc_sdu));
+	sdu = g_malloc0(sizeof(struct rlc_sdu));
 	return sdu;
 }
 
@@ -418,7 +418,7 @@ rlc_frag_create(tvbuff_t *tvb, enum rlc_mode mode, packet_info *pinfo,
 {
 	struct rlc_frag *frag;
 
-	frag = se_alloc0(sizeof(struct rlc_frag));
+	frag = g_malloc0(sizeof(struct rlc_frag));
 	rlc_frag_assign(frag, mode, pinfo, seq, li);
 	rlc_frag_assign_data(frag, tvb, offset, length);
 
@@ -461,6 +461,7 @@ free_sequence_table_entry_data(gpointer data)
 		g_list_free(list->list);
 		list->list = NULL;   /* for good measure */
 	}
+        g_free(list);
 }
 
 static void
@@ -872,7 +873,7 @@ rlc_is_duplicate(enum rlc_mode mode, packet_info *pinfo, guint16 seq,
 	list = g_hash_table_lookup(sequence_table, &lookup.ch);
 	if (!list) {
 		/* we see this channel for the first time */
-		list = se_alloc0(sizeof(*list));
+		list = g_malloc0(sizeof(*list));
 		rlc_channel_assign(&list->ch, mode, pinfo);
 		g_hash_table_insert(sequence_table, &list->ch, list);
 	}
-- 
1.7.10.4

